### PR TITLE
rpc: Add rpcallowstop

### PR DESCRIPTION
**Pull request by cvengler**
This adds a feature to disable to ability to stop Bitcoin Core over the JSON RPC interface except if the sender is localhost.
Based on a suggestion from hebasto in #16543.

All in all it adds a new parameter called `rpcallowstop` which is by default set to 1. If the parameter is set to 0 it will block all RPC requests with stop Bitcoin Core except if the sender is localhost.

I tested it with the following config file:
```
regtest=1

[regtest]
server=1
rpcallowstop=0
rpcport=8332
rpcuser=foo
rpcpassword=bar
rpcallowip=192.168.178.71
rpcallowip=127.0.0.1
rpcbind=192.168.178.71
rpcbind=127.0.0.1
```
A curl request using the internal ip would result in an error. However using 127.0.0.1 would work.

### CODE REVIEW

**Review by marpme — CHANGES_REQUESTED** [review]
I suggest changing the default value to: `false`. 

As it opens an attack vector to stop all core nodes remotely and additionally previously none of the users would expect that remote servers could stop their nodes.

File: src/rpc/server.cpp [hunk 1]
```diff
@@ -169,11 +170,39 @@ UniValue stop(const JSONRPCRequest& jsonRequest)
             }.ToString());
     // Event loop will exit after current HTTP requests have been handled, so
     // this reply will get back to the client.
-    StartShutdown();
-    if (jsonRequest.params[0].isNum()) {
-        MilliSleep(jsonRequest.params[0].get_int());
+    if(rpcallowstop)
```
Comment (hebasto) [inline src/rpc/server.cpp]: Could you follow [Coding Style (C++)](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#coding-style-c)?

**Review by hebasto — CHANGES_REQUESTED** [review]
Concept ACK

File: src/rpc/server.cpp [hunk 2]
```diff
@@ -169,11 +170,39 @@ UniValue stop(const JSONRPCRequest& jsonRequest)
...
+        {
+            if(jsonRequest.peerAddr[i] == ':')
+            {
+                break;
+            }
+            ip += jsonRequest.peerAddr[i];
+        }
+        if(ip == "127.0.0.1")
+        {
+            StartShutdown();
```
Comment (jonatack) [inline src/rpc/server.cpp]: This block is redundant with the one at L174 and could be extracted.

File: src/init.cpp [hunk 3]
```diff
@@ -517,6 +517,7 @@ void SetupServerArgs()
 
     gArgs.AddArg("-rest", strprintf("Accept public REST requests (default: %u)", DEFAULT_REST_ENABLE), ArgsManager::ALLOW_ANY, OptionsCategory::RPC);
     gArgs.AddArg("-rpcallowip=<ip>", "Allow JSON-RPC connections from specified source. Valid for <ip> are a single IP (e.g. 1.2.3.4), a network/netmask (e.g. 1.2.3.4/255.255.255.0) or a network/CIDR (e.g. 1.2.3.4/24). This option can be specified multiple times", ArgsManager::ALLOW_ANY, OptionsCategory::RPC);
+    gArgs.AddArg("-rpcallowstop", "Allow stop over a remote RPC client", ArgsManager::ALLOW_ANY, OptionsCategory::RPC);
```
Comment (jonatack) [inline src/init.cpp]: Suggest: "Accept shutdown (stop) commands from remote RPC clients."

**Review by jonatack — COMMENTED** [review]
Needs test coverage. EDIT:

> #12763 is a more general solution

Agreed.

**Comment by promag** [comment]
> I suggest changing the default value to: `false`.
> 
> As it opens an attack vector to stop all core nodes remotely and additionally previously none of the users would expect that remote servers could stop their nodes.

@marpme that would be a breaking change. You should always restrict RPC access anyway.

**Comment by promag** [comment]
Maybe support any RPC, like `-rpcrestrict=senttoaddress,sendmany,stop`?

**Comment by ryanofsky** [comment]
#12763 is a more general solution

**Comment by emilengler** [comment]
@marpme 
I'm skeptical about this. Maybe some people are relying on a remote stop. I think your suggestion could be added later if the PR gets into the code

File: src/init.cpp (hunk 3 above)
Comment (emilengler) [reply src/init.cpp]: I also thought about this but I want that it is clear that it is referring to the "stop" command.

File: src/rpc/server.cpp (hunk 2 above)
Comment (emilengler) [reply src/rpc/server.cpp]: Are you sure about this?
This block should only be executed if remote stop is disabled so a separate string slicing even if it isn't necessary is a bit too much overhead imo.

File: src/rpc/server.cpp (hunk 1 above)
Comment (emilengler) [reply src/rpc/server.cpp]: Sure thing

Comment (emilengler) [reply src/rpc/server.cpp]: Done

Comment (hebasto) [reply src/rpc/server.cpp]:
```
$ git diff -U0 HEAD~2.. src/rpc/server.cpp | ./contrib/devtools/clang-format-diff.py -p1 -v
--- src/rpc/server.cpp	(before formatting)
+++ src/rpc/server.cpp	(after formatting)
@@ -170,36 +170,28 @@
             }.ToString());
     // Event loop will exit after current HTTP requests have been handled, so
     // this reply will get back to the client.
-    if (rpcallowstop)
-    {
+    if (rpcallowstop) {
         StartShutdown();
         if (jsonRequest.params[0].isNum()) {
             MilliSleep(jsonRequest.params[0].get_int());
         }
         return "Bitcoin server stopping";
-    }
-    else
-    {
+    } else {
         // Check if IP is 127.0.0.1
         std::string ip;
-        for (int i = 0; i < jsonRequest.peerAddr.length(); i++)
-        {
-            if (jsonRequest.peerAddr[i] == ':')
-            {
+        for (int i = 0; i < jsonRequest.peerAddr.length(); i++) {
+            if (jsonRequest.peerAddr[i] == ':') {
                 break;
             }
             ip += jsonRequest.peerAddr[i];
         }
-        if (ip == "127.0.0.1")
-        {
+        if (ip == "127.0.0.1") {
             StartShutdown();
             if (jsonRequest.params[0].isNum()) {
                 MilliSleep(jsonRequest.params[0].get_int());
             }
             return "Bitcoin server stopping";
-        }
-        else
-        {
+        } else {
             throw JSONRPCError(RPC_FORBIDDEN_BY_SAFE_MODE, "You are not allowed to stop");
         }
     }
```

Comment (hebasto) [reply src/rpc/server.cpp]: > You can use the provided [clang-format-diff script](https://github.com/bitcoin/bitcoin/blob/master/contrib/devtools/README.md#clang-format-diffpy) tool to clean up patches automatically before submission.

File: src/init.cpp [hunk 4]
```diff
@@ -749,7 +750,8 @@ static bool AppInitServers()
     RPCServer::OnStopped(&OnRPCStopped);
     if (!InitHTTPServer())
         return false;
-    StartRPC();
+    if (gArgs.GetBoolArg("-rpcallowstop", true)) StartRPC();
+    else StartRPC(false);
```
Comment (hebasto) [inline src/init.cpp]: `if ... else ...` is confusing.
Is `then` missed or it should be instead of `else`?

Comment (hebasto) [inline src/init.cpp]: Just `StartRPC(gArgs.GetBoolArg("-rpcallowstop", true));` ?

Comment (emilengler) [reply src/init.cpp]: Yes

Comment (emilengler) [reply src/init.cpp]: Fixed

**Comment by luke-jr** [comment]
Yeah, concept NACK on something this specific.

**Comment by laanwj** [comment]
Tend towards NACK, seems awfully specific. I'd prefer like to avoid wild growth of features such as this.

Bitcoin core RPC is *not* safe for use remotely in any case, definitely not with untrusted users. Blocking one "attack vector" leaves tons of DoS and potential other vulnerabilities.

**Comment by emilengler** [comment]
Ok I will close this now because of all the arguments above especially that it is too specific, feel free to steal if you need this in future :)

**Pull request closed without merging**